src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -1082,24 +1070,51 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     }
 
     pub fn incremental_args(&self, unit: &Unit) -> CargoResult<Vec<String>> {
-        if self.incremental_enabled {
-            if unit.pkg.package_id().source_id().is_path() {
-                // Only enable incremental compilation for sources the user can modify.
-                // For things that change infrequently, non-incremental builds yield
-                // better performance.
-                // (see also https://github.com/rust-lang/cargo/issues/3972)
-                return Ok(vec![format!("-Zincremental={}",
-                                       self.layout(unit.kind).incremental().display())]);
-            } else if unit.profile.codegen_units.is_none() {
-                // For non-incremental builds we set a higher number of
-                // codegen units so we get faster compiles. It's OK to do
-                // so because the user has already opted into slower
-                // runtime code by setting CARGO_INCREMENTAL.
-                return Ok(vec![format!("-Ccodegen-units={}", ::num_cpus::get())]);
-            }
+        // There's a number of ways to configure incremental compilation right
+        // now. In order of descending priority (first is highest priority) we
+        // have:
+        //
+        // * `CARGO_INCREMENTAL` - this is blanket used unconditionally to turn
+        //   on/off incremental compilation for any cargo subcommand. We'll
+        //   respect this if set.
+        // * `build.incremental` - in `.cargo/config` this blanket key can
+        //   globally for a system configure whether incremental compilation is
+        //   enabled. Note that setting this to `true` will not actually affect
+        //   all builds though. For example a `true` value doesn't enable
+        //   release incremental builds, only dev incremental builds. This can
+        //   be useful to globally disable incremental compilation like
+        //   `CARGO_INCREMENTAL`.
+        // * `profile.dev.incremental` - in `Cargo.toml` specific profiles can
+        //   be configured to enable/disable incremental compilation. This can
+        //   be primarily used to disable incremental when buggy for a project.
+        // * Finally, each profile has a default for whether it will enable
+        //   incremental compilation or not. Primarily development profiles
+        //   have it enabled by default while release profiles have it disabled
+        //   by default.
+        let global_cfg = self.config.get_bool("build.incremental")?.map(|c| c.val);
+        let profile = match unit.profile.incremental {
+            Incremental::Explicit(v) => v,
+            Incremental::Default(default) => default && global_cfg.unwrap_or(default),
```

Comment:
Ok, I am staring at this boolean logic for like ten minutes, I can't persuade myself if it is right or wrong, could you recheck it just in case? :) 

Here are my doubts:

I believe `&& global_cfg.unwrap_or(default)` condition is not not necessary here. That is, if you remove it, the code would behave exactly the same.

1) if global cfg is not set, then `default && global_cfg.unpwrap_or(default) == default && default == default`
2) if global cfg is true, then `default && global_cfg.unwrap_or(default) == default && true == default`.
3) if global cfg is false, than we'll bail out on the next like, without inspecting `profile` at all.


>!self.incremental_env.unwrap_or(global_cfg.unwrap_or(profile))

So, if global_cfg is true, we do incremental regradless of profile setting? This is not what the doc comment says?

Reply:
Oh oops, good catch! 

The `default` here was intended to affect `global_cfg`, ensuring that if the default is false and the global cfg says true, we still turn it off. I'll update